### 爬虫的基本原理
* HTML决定了网站的原始容貌，其中的信息可以修改，但是并不会将这些修改返回给服务器，也就是下次打开的时候本地的修改就会消失

### 抓包，_抓取数据包_
* 在网页上操作时，拦截传输过来的数据包，在分析其中包含的信息
* 工具**MAC charles**


In [2]:
import requests
from bs4 import BeautifulSoup

In [75]:
# 尝试获取某页面，并获得一个response对象
r = requests.get('http://gitbook.cn/')

In [77]:
# text中包含所有的html页面代码
# r.text

___
## 图片爬取
* https://unsplash.com/

___

# 反扒教学

## 常见反扒技术

* **User-Agent**: 网站通过验证接收到的 User-Agent 信息中是否包含 python 字样来拒绝访问呢
* **Referer**: 表示请求从哪发出的。网站可以指定从某个地址跳转进来，对于其他的地址一律禁止访问
* **authorization**: 网站可以禁止没有用户ID的访问，我们那可以在header中添加client-ID来解决这个问题
* **添加延迟**:通过用time.sleep(1)是脚本休息1秒的方式来防止反爬虫
* ****:
* ****:
* ****:
* ****:
* ****:


In [88]:
import requests
import json

target = 'http://unsplash.com/napi/feeds/home'

headers = {'authorization':'01234'}

req = requests.get(url=target, verify=False)

#print(req.text)
html = json.loads(req.text)
next_page = html['next_page']

next_page

/Users/jasonmei/Anaconda/anaconda/envs/OSMNX/lib/python3.6/site-packages/urllib3/connectionpool.py:858: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  InsecureRequestWarning)


'https://api.unsplash.com/feeds/home?after=7bf36780-cf76-11e8-8080-80001048bbcd'

In [89]:
for each in html['photos']:

    print('图片ID:',each['id'])

图片ID: e8SKBWrwoB8
图片ID: lQoC90sqn8s
图片ID: yQAtbM9Scwc
图片ID: RW8nmjS04B8
图片ID: Hf2Px71VRb4
图片ID: FVRTLKgQ700
图片ID: 3yaTFzlnJhs
图片ID: xUKfj5AE0O4
图片ID: 2YdDsm5T6Eo
图片ID: F-JUGSYNjZU


In [80]:
r = requests.get('https://unsplash.com/')

In [81]:
html = BeautifulSoup(r.text)

In [82]:
html

<!DOCTYPE html>
<html lang="en"><head><meta content="IE=Edge" http-equiv="X-UA-Compatible"/><link href="https://images.unsplash.com" rel="preconnect"/><link href="https://stats.g.doubleclick.net" rel="preconnect"/><link href="/a/opensearch.xml" rel="search" type="application/opensearchdescription+xml"/><style data-href="/a/main.26533.css">
._3GpgA{font-family:-apple-system,BlinkMacSystemFont,San Francisco,Helvetica Neue,Helvetica,Ubuntu,Roboto,Noto,Segoe UI,Arial,sans-serif}._3Ysek{margin:0 0 16px;font-size:46px;line-height:1.2}@media (min-width:768px){._3Ysek{margin-bottom:24px;font-size:72px}}._2RmO0{margin:0 0 16px;font-size:28px;line-height:1.25}@media (min-width:768px){._2RmO0{font-size:46px;line-height:1.2}}.TBVsU{margin:0 0 16px;font-size:21px;line-height:1.2}@media (min-width:768px){.TBVsU{font-size:28px}}._2uOAz{margin:0;font-size:18px;line-height:1.34}._1iWCF{font-size:15px}@media (min-width:768px){._1iWCF{font-size:18px}}._22v5L{line-height:1}.U8wGh{line-height:1.2}._13hUA{l

In [76]:
div = html.find_all('div', class_ = '_3aQBo _2T3hc _2cfKf')

In [77]:
div

[<div class="_3aQBo _2T3hc _2cfKf"><div class="_1pn7R"><div><figure itemprop="image" itemscope="" itemtype="http://schema.org/ImageObject"><div class="yVU8k"><a class="_2Mc8_" href="/photos/B2bzZZHLBg4" itemprop="contentUrl" title="View the photo by Ryoji Iwata"><div class="_5nYwn IEpfq" style="padding-bottom:150.009%"><img class="_2zEKz" data-test="standard-photo-grid-multi-col-img" itemprop="thumbnailUrl" sizes="(min-width: 1335px) 416px, (min-width: 992px) calc(calc(100vw - 72px) / 3), (min-width: 768px) calc(calc(100vw - 48px) / 2), 100vw" src="https://images.unsplash.com/photo-1539443303960-a90a4e7ecc97?ixlib=rb-0.3.5&amp;ixid=eyJhcHBfaWQiOjEyMDd9&amp;s=a6c2af73dc3866b58f5de0735502f627&amp;w=1000&amp;q=80" srcset="https://images.unsplash.com/photo-1539443303960-a90a4e7ecc97?ixlib=rb-0.3.5&amp;ixid=eyJhcHBfaWQiOjEyMDd9&amp;s=a6c2af73dc3866b58f5de0735502f627&amp;auto=format&amp;fit=crop&amp;w=100&amp;q=60 100w, https://images.unsplash.com/photo-1539443303960-a90a4e7ecc97?ixlib=rb-0.

In [79]:
BeautifulSoup(str(div[0]))

<html><head></head><body><div class="_3aQBo _2T3hc _2cfKf"><div class="_1pn7R"><div><figure itemprop="image" itemscope="" itemtype="http://schema.org/ImageObject"><div class="yVU8k"><a class="_2Mc8_" href="/photos/B2bzZZHLBg4" itemprop="contentUrl" title="View the photo by Ryoji Iwata"><div class="_5nYwn IEpfq" style="padding-bottom:150.009%"><img class="_2zEKz" data-test="standard-photo-grid-multi-col-img" itemprop="thumbnailUrl" sizes="(min-width: 1335px) 416px, (min-width: 992px) calc(calc(100vw - 72px) / 3), (min-width: 768px) calc(calc(100vw - 48px) / 2), 100vw" src="https://images.unsplash.com/photo-1539443303960-a90a4e7ecc97?ixlib=rb-0.3.5&amp;ixid=eyJhcHBfaWQiOjEyMDd9&amp;s=a6c2af73dc3866b58f5de0735502f627&amp;w=1000&amp;q=80" srcset="https://images.unsplash.com/photo-1539443303960-a90a4e7ecc97?ixlib=rb-0.3.5&amp;ixid=eyJhcHBfaWQiOjEyMDd9&amp;s=a6c2af73dc3866b58f5de0735502f627&amp;auto=format&amp;fit=crop&amp;w=100&amp;q=60 100w, https://images.unsplash.com/photo-1539443303960-

In [39]:
for chapter in a:
    print(chapter.string, server + chapter.get('href'))

第1314章 你的选择（终） https://www.biqukan.com//1_1094/17967679.html
第1313章 青灯古庙伴一生 https://www.biqukan.com//1_1094/17967303.html
第1312章 左岸 https://www.biqukan.com//1_1094/17964263.html
第1311章 一念，永恒 https://www.biqukan.com//1_1094/17956049.html
第1310章 念起天劫现 https://www.biqukan.com//1_1094/17953054.html
第1309章 见神 https://www.biqukan.com//1_1094/17942081.html
第1308章 引导与开启 https://www.biqukan.com//1_1094/17939510.html
第1307章 起源 https://www.biqukan.com//1_1094/17930894.html
第1306章 罗天无情 https://www.biqukan.com//1_1094/17927946.html
第1305章 逆陨 https://www.biqukan.com//1_1094/17918892.html
第1304章 罗天指！ https://www.biqukan.com//1_1094/17916686.html
第1303章 惨烈 https://www.biqukan.com//1_1094/17908165.html
外传1 柯父。 https://www.biqukan.com//1_1094/5386269.html
外传2 楚玉嫣。 https://www.biqukan.com//1_1094/5386270.html
外传3 鹦鹉与皮冻。 https://www.biqukan.com//1_1094/5386271.html
第一章 他叫白小纯 https://www.biqukan.com//1_1094/5403177.html
第二章 火灶房 https://www.biqukan.com//1_1094/5428081.html
第三章 六句真言 https://www.biqukan.com//

第801章 你是故意的吧 https://www.biqukan.com//1_1094/14491022.html
第802章 大天师召见 https://www.biqukan.com//1_1094/14497478.html
第803章 和我斗？ https://www.biqukan.com//1_1094/14499182.html
第804章 群起攻之 https://www.biqukan.com//1_1094/14505827.html
第805章 殿内炼火！ https://www.biqukan.com//1_1094/14507639.html
第806章 震撼公侯！ https://www.biqukan.com//1_1094/14513773.html
第807章 地品雷火劫 https://www.biqukan.com//1_1094/14515429.html
第808章 有本事再来一个 https://www.biqukan.com//1_1094/14521505.html
第809章 我们都是师字辈 https://www.biqukan.com//1_1094/14526069.html
第810章 我舍不得你们啊 https://www.biqukan.com//1_1094/14532305.html
第811章 宠物…… https://www.biqukan.com//1_1094/14533946.html
第812章 那里……有人！ https://www.biqukan.com//1_1094/14540019.html
第813章 守陵老爷爷…… https://www.biqukan.com//1_1094/14541795.html
第814章 有问题，找老爷爷 https://www.biqukan.com//1_1094/14548050.html
第815章 突飞猛进 https://www.biqukan.com//1_1094/14552038.html
第816章 十九色火 https://www.biqukan.com//1_1094/14559005.html
第817章 白小纯？！ https://www.biqukan.com//1_1094/14561057.html
第818章

In [16]:
print(text[0].text.replace('\xa0'*8, '\n\n'))

##  上面这行代码中：
##  .text，用于去除 br 标签
##  .replace('\xa0'*8, '\n\n')，用于将八个空格符合 '\xa0'*8 以回车 '\n\n' 代替



帽儿山，位于东林山脉中，山下有一个村子，民风淳朴，以耕田为生，与世隔绝。



清晨，村庄的大门前，整个村子里的乡亲，正为一个十五六岁少年送别，这少年瘦弱，但却白白净净，看起来很是乖巧，衣着尽管是寻常的青衫，可却洗的泛白，穿在这少年的身上，与他目中的纯净搭配在一起，透出一股子灵动。



他叫白小纯。



“父老乡亲们，我要去修仙了，可我舍不得你们啊。”少年满脸不舍，原本就乖巧的样子，此刻看起来更为纯朴。



四周的乡亲，面面相觑，顿时摆出难舍之色。



“小纯，你爹娘走的早，你是个……好孩子！！难道你不想长生了么，成为仙人就可以长生，能活的很久很久，走吧，雏鹰长大，总有飞出去的那一天。”人群内走出一个头发花白的老者，说道好孩子三个字时，他顿了一下。



“在外面遇到任何事情，都要坚持下去，走出村子，就不要回来，因为你的路在前方！”老人神色慈祥，拍了拍少年的肩膀。



“长生……”白小纯身体一震，目中慢慢坚定起来，在老者以及四周乡亲鼓励的目光下，他重重的点了点头，深深的看了一眼四周的乡亲，转身迈着大步，渐渐走出了村子。



眼看少年的身影远去，村中的众人，一个个都激动起来，目中的难舍刹那就被喜悦代替，那之前满脸慈祥的老者，此刻也在颤抖，眼中流下泪水。



“苍天有眼，这白鼠狼，他终于……终于走了，是谁告诉他在附近看到仙人的，你为村子立下了大功！”



“这白鼠狼终于肯离开了，可怜我家的几只鸡，就因为这白鼠狼怕鸡打鸣，不知用了什么方法，唆使一群孩子吃鸡肉，把全村的鸡都给吃的干干净净……”



“今天过年了！”欢呼之声，立刻在这不大的村子里，沸腾而起，甚至有人拿出了锣鼓，高兴的敲打起来。



村子外，白小纯还没等走远，他就听到了身后村子内，传出了敲锣打鼓的声音，还夹着欢呼。



白小纯脚步一顿，神色有些古怪，干咳一声，伴随着耳边传来的锣鼓，白小纯顺着山路，走上了帽儿山。



这帽儿山虽不高，却灌木杂多，虽是清晨，可看起来也是黑压压一片，很是安静。



“听二狗说，他前几天在这里被一头野猪追赶时，看到天上有仙人飞过……”白小纯走在山路上，心脏怦怦跳动时，忽然一旁的灌林中传来阵阵哗哗声，似野猪一样，这声音来的突然，让本就紧张的白小纯，顿时背后发凉。



“谁，谁在那里！”白小纯右手快速从行囊中拿出四把斧头，六把柴刀，还觉得不放心，又从怀里取出

____

In [54]:
k_ls, v_ls = [], []
for key in r.headers.keys():
    k_ls.append(key)
for value in r.headers.values():
    v_ls.append(value)
for i,_ in enumerate(k_ls):
    print(k_ls[i],': ', v_ls[i])

Server :  nginx/1.6.1
Date :  Sun, 14 Oct 2018 03:45:57 GMT
Content-Type :  text/html
Transfer-Encoding :  chunked
Connection :  keep-alive
Vary :  Accept-Encoding
Cache-Control :  no-cache, must-revalidate
Expires :  Sat, 26 Jul 1997 05:00:00 GMT
Pragma :  no-cache, no-cache
DPOOL_HEADER :  dryad62
Content-Encoding :  gzip
SINA-LB :  aGEuMTY4LmcxLmh5ZHMubGIuc2luYW5vZGUuY29t
SINA-TS :  Y2ZjYTk0Y2UgMCA0IDQgNSA0MQo=


In [59]:
import requests

# 关键字
url = 'https://image.baidu.com/search/index?tn=baiduimage&ipn=r&ct=201326592&cl=2&lm=-1&st=-1&fm=result&fr=&sf=1&fmq=1539490550614_R&pv=&ic=0&nc=1&z=&se=1&showtab=0&fb=0&width=&height=&face=0&istype=2&ie=utf-8'
keyword = "刘德华"

try:
    kv = {"word":keyword}
    r = requests.get(url, params = kv)
    print(r.request.url)
    r.raise_for_status()
    print(len(r.text))
    #print(r.text)
except:
    print("爬取失败")

https://image.baidu.com/search/index?tn=baiduimage&ipn=r&ct=201326592&cl=2&lm=-1&st=-1&fm=result&fr=&sf=1&fmq=1539490550614_R&pv=&ic=0&nc=1&z=&se=1&showtab=0&fb=0&width=&height=&face=0&istype=2&ie=utf-8&word=%E5%88%98%E5%BE%B7%E5%8D%8E
117982


In [65]:
import requests
import os

url = "http://image.nationalgeographic.com.cn/2015/0121/20150121033625957.jpg"
root = "Images/"
path = root + url.split('/')[-1]

try:
    if not os.path.exists(root):
        os.mkdir(root)
    if not os.path.exists(path):
        r = requests.get(url)
        with open(path,'wb') as f:
            f.write(r.content)
            f.close()
            print("文件保存成功")
    else :
        print("文件已存在")
except:
    print("爬取失败")


文件保存成功


#### 为URL的参数部分传递参数

```python
para_dict = {'key1':"v1", "key2":"v2"}
r = requests.get("http://httpbin.org/get", params=para_dict)

"""以上相当于，/get?key1=v1&key2=v2"""
```

#### 响应内容
```python
r = requests.get('https://api.github.com/events')
r.text
```

#### 
```python
```

#### 
```python
```

#### 
```python
```

#### 
```python
```

#### 
```python
```

___
<font color=blue>
## HTTP
</font>
### 基本概念
* HTTP协议是用于**从万维网服务器传输超文本到客户端（本地浏览器）的传送协议**
* 基于 TCP/IP 协议来传递数据
* 客户端**通过URL向服务器发送所有请求**
* 无连接（服务器限制每次连接只处理一个请求，处理完后，即刻断开连接）
* 媒体独立（任何类型的数据都可通过HTTP发送）
* 无状态（服务器对于事务处理没有记忆能力）

> ### URL
* http://www.aspxfans.com:8080/news/index.asp?boardID=5&ID=24618&page=1#name
* "http:": 说明网页用的是HTTP协议
* "www.aspxfans.com": 域名
* ":8080": 端口
* "/news/": 虚拟目录部分
* "index.asp": 文件名部分, 位于最后一个 "/" 和 "?" 之间
* "boardID=5&ID=24618&page=1": 参数（搜索）部分, 位于 "?" 和 "#" 之间
* "name": M锚部分, 从 "#" 开始到最后
* "":
* "":
* "":
* "":
* "":


### HTTP请求
1. get: 请求指定的页面信息
2. head: 请求指定的页面的报头
3. post: 向指定资源提交数据进行处理请求（例如上传文件，提交表单）。post可以更改现有资源，以及建立新的资源
4. put: 取代服务器内指定文档的内容
5. delete: 请求服务器删除指定页面
6. options: 运行客户端查看服务器的性能